modules/standard/Heap.chpl

```diff
@@ -0,0 +1,420 @@
+    */
+    pragma "no doc"
+    proc _greater(x: eltType, y: eltType) {
+      return chpl_compare(x, y, comparator) > 0;
+    }
+
+    /*
+      helper procs to maintain the heap
+    */
+    pragma "no doc"
+    proc _heapify_up(in pos: int) {
+      while (pos) {
+        var parent = pos / 2;
+        if (_greater(_data[pos],_data[parent])) {
+          _data[parent] <=> _data[pos];
+          pos = parent;
+        }
+        else break;
+      }
+    }
+
+    pragma "no doc"
+    proc _heapify_down(in pos: int) {
+      while (pos < _data.size) {
+        // find the child node with greater value
+        var greaterChild = pos*2;
+        if (greaterChild >= _data.size) then break; // reach leaf node, break
+        if (greaterChild + 1 < _data.size) {
+          // if the right child node exists
+          if (_greater(_data[greaterChild+1],_data[greaterChild])) {
+            // if the right child is greater, then update the greaterChild
+            greaterChild += 1;
+          }
+        }
+        // if the greaterChild's value is greater than current node, then swap and continue
+        if (_greater(_data[greaterChild],_data[pos])) {
+          _data[greaterChild] <=> _data[pos];
+          pos = greaterChild;
+        }
+        else break;
+      }
+    }
+
+    pragma "no doc"
+    proc _push(in element: eltType)
+    lifetime this < element {
+      _data.append(element);
+      _heapify_up(_data.size-1);
+    }
+    /*
+      Push an element into the heap
+
+      :arg element: The element that will be pushed
+      :type element: `eltType`
+    */
+    proc push(in element: eltType)
+    lifetime this < element {
+      _enter();
+      _push(element);
+      _leave();
+    }
+
+    /*
+      Push elements of a list into the heap
+
+      :arg x: The list of which elements is to push.
+      :type x: `list(eltType)`
+    */
+    proc push(ref x: list(eltType)) {
+      _enter();
+      for e in x do
+        _push(x);
+      _leave();
+    }
+
+    /*
+      Push elements in an array into a heap.
+
+      :arg x: The array of which elements is to push.
+      :type x: `[?d] eltType`
+    */
+    proc push(ref x:[?d] eltType) {
+      _enter();
+      for e in x do
+        _push(e);
+      _leave();
+    }
+
+    /*
+      Pop an element and return it.
+
+      :return: the top element
+      :rtype: eltType
+    */
+    proc pop(): eltType {
+      _enter();
+      if (boundsChecking && isEmpty()) {
+        boundsCheckHalt("Called \"heap.pop\" on an empty heap.");
+      }
+
+      if _data.size != 1 then
+        _data(0) <=> _data(_data.size-1);
+
+      var ret = _data.pop();
+      _heapify_down(0);
+      _leave();
+      return ret;
+    }
+
+    /*
+      Iterate over the elements of this heap
+
+        .. note::
+          Elements are not yielded in order
+
+    */
+    iter these() ref {
+      for e in _data {
+        yield e;
+      }
+    }
+
+    /*
+      Iterate over the elements of this heap
```

Comment:
It may be implicit in the name of the method 'consume', but I think the documentation should note that the elements yielded are removed from the heap.

Reply:
https://github.com/chapel-lang/chapel/pull/15997/commits/b1cc51c5baaa5b4feed46431d5ae75c4368f95bd

Reply:
https://github.com/chapel-lang/chapel/pull/15997/commits/b1cc51c5baaa5b4feed46431d5ae75c4368f95bd A note added here